tests/core/test_agent.py

```diff
@@ -292,6 +292,63 @@ def test_two_stage_fallback_without_deny_suggestion(
     assert "The intent 'out_of_scope' must be present" in str(execinfo.value)
 
 
+@pytest.mark.parametrize(
+    "domain, policy_config",
+    [
+        (
+            {"actions": ["other-action"]},
+            {
+                "policies": [
+                    {"name": "RulePolicy", "core_fallback_action_name": "my_fallback"}
+                ]
+            },
+        )
+    ],
+)
+def test_rule_policy_without_fallback_action_present(
+    domain: Dict[Text, Any], policy_config: Dict[Text, Any]
+):
+    with pytest.raises(InvalidDomain):
+        Agent(
+            domain=Domain.from_dict(domain),
+            policies=PolicyEnsemble.from_dict(policy_config),
+        )
```

Comment:
To catch wrong error message formatting like above 😉 
```suggestion
    with pytest.raises(InvalidDomain) as execinfo:
        Agent(
            domain=Domain.from_dict(domain),
            policies=PolicyEnsemble.from_dict(policy_config),
        )
    assert "fallback action 'my_fallback' must be present" in str(execinfo.value)
```

Reply:
I am personally not a huge fan of testing the content of a log message / exception message and prefer to test **that** the exception / log message is thrown / printed. The reason is that it makes the test very fragile to minor changes and the message content not really belongs to the functionality for me.

I also think that above are somewhat weak arguments and tried to find some best practices:
- https://softwareengineering.stackexchange.com/questions/384164/should-log-statements-be-tested
- https://stackoverflow.com/questions/11998713/should-we-unit-test-logging
- https://softwareengineering.stackexchange.com/questions/406079/testing-the-wording-of-an-exception-message
- https://thecodebuzz.com/unit-testing-exceptions-best-practices-net-core/ (this one is somewhat pro)

I'd say that it's somewhat optional to test it and as the correct formatting of the message is not critical, it's okay to just test **that** the exception happens.

Wdyt?

Reply:
Thanks for the resources!

Seems to me that there are two good ways:
- only test that it is thrown
- also test that the correct error is thrown (there are several cases where our code raises `InvalidDomain`, for different reasons with different messages)
If we want to test if it is the correct error, we can reduce the snippet to be only `"my_fallback"` to make it less fragile. The name of the fallback action should always be part of the message even if we decide to rephrase it.

Both options would be fine with me, feel free to pick one.
We should quickly go through the rest of the test file for consistency though. `test_form_without_form_policy`, `test_trigger_without_mapping_policy`, `test_two_stage_fallback_without_deny_suggestion` and `test_pull_model_with_invalid_domain` all test for message snippets.

Reply:
I agree. Testing the exception message now to see that the exception was caused by the RulePolicy 👍